app/javascript/history/history.jsx

```diff
@@ -1,213 +1,106 @@
+import { ItemListItem } from '../src/components/ItemList/ItemListItem';
 
 export class History extends Component {
   constructor(props) {
     super(props);
 
     const { availableTags } = this.props;
-    this.state = {
-      query: '',
-      items: [],
-      totalCount: 0,
-      index: null,
-      itemsLoaded: false,
-      hitsPerPage: 100,
-      availableTags,
-      selectedTags: [],
-      page: 0,
-      showNextPageButton: false,
-    };
+    this.state = defaultState({ availableTags });
 
-    this.handleTyping = debounce(this.handleTyping.bind(this), 300, {
+    // bind and initialize all shared functions
+    this.onSearchBoxType = debounce(onSearchBoxType.bind(this), 300, {
       leading: true,
     });
+    this.loadNextPage = loadNextPage.bind(this);
+    this.performInitialSearch = performInitialSearch.bind(this);
+    this.search = search.bind(this);
+    this.toggleTag = toggleTag.bind(this);
   }
 
   componentDidMount() {
-    const index = setupAlgoliaIndex({
-      containerId: 'history',
-      indexName: 'UserHistory',
-    });
-
-    // get default result set from Algolia
     const { hitsPerPage } = this.state;
-    index.search('', { hitsPerPage }).then(content => {
-      this.setState({
-        items: content.hits,
-        totalCount: content.nbHits,
-        index,
-        itemsLoaded: true,
-        showNextPageButton: content.hits.length === hitsPerPage,
-      });
-    });
-  }
-
-  handleTyping = event => {
-    const query = event.target.value;
-    const { selectedTags } = this.state;
-
-    this.setState({ page: 0, items: [] });
-    this.search(query, { tags: selectedTags });
-  };
 
-  toggleTag = (event, tag) => {
-    event.preventDefault();
-
-    const { query, selectedTags } = this.state;
-    const newTags = selectedTags;
-    if (newTags.indexOf(tag) === -1) {
-      newTags.push(tag);
-    } else {
-      newTags.splice(newTags.indexOf(tag), 1);
-    }
-
-    this.setState({ selectedTags: newTags, page: 0, items: [] });
-    this.search(query, { tags: newTags });
-  };
-
-  loadNextPage = () => {
-    const { query, selectedTags, page } = this.state;
-    this.setState({ page: page + 1 });
-    this.search(query, { selectedTags });
-  };
-
-  search(query, { tags }) {
-    const { index, hitsPerPage, page, items } = this.state;
-    const filters = { hitsPerPage, page };
-
-    if (tags && tags.length > 0) {
-      filters.tagFilters = tags;
-    }
-
-    index.search(query, filters).then(content => {
-      const allItems = [...items, ...content.hits];
-
-      this.setState({
-        query,
-        items: allItems,
-        totalCount: content.nbHits,
-        showNextPageButton: content.hits.length === hitsPerPage,
-      });
+    this.performInitialSearch({
+      containerId: 'history',
+      indexName: 'UserHistory',
+      searchOptions: {
+        hitsPerPage,
+      },
     });
   }
 
-  renderNoItems() {
+  renderEmptyItems() {
```

Comment:
This part is duplicated among the two components, I thought it would have complicated things to extract, because the "empty" state for the two pages is slightly different. Codeclimate complains about that obviously :D